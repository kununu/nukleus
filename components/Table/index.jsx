import React, {PropTypes} from 'react';

import styles from './index.scss';

function getRows (cols) {
  return (
    [...Array(cols[0].length)].reduce((acc, item, index) => {
      acc[index] = [];
      cols.forEach(col => {
        acc[index].push(col[index]);
      });
      return acc;
    }, [])
  );
}

function getRow (value, index) {
  if (!value) return <td key={index} />;
  return (
    <td key={index}>
      {value}
    </td>);
}

function getTitles (value, index) {
  return !value ?
    <th key={index} className={styles.emptyHeader} /> : // eslint-disable-line react/no-array-index-key
    <th key={index}><span>{value}</span></th>; // eslint-disable-line react/no-array-index-key
}

export default function Table ({
  items,
  defaultTitleCount,
  defaultRowCount,
  tableStyle
}) {
  const keys = Object.keys(items);
  const hasItems = keys.length !== 0;
  const titles = hasItems ?
                    keys :
                    [...Array(defaultTitleCount)];
  const rowCount = hasItems ?
                    items[keys[0]].length :
                    defaultRowCount;
  const rows = hasItems ?
                getRows(keys.map(item => items[item])) :
                [...Array(defaultTitleCount)];

  return (
    <div className={`${styles.tableContainer} ${styles[tableStyle]}`}>
      <table className={styles.table}>
        <thead>
          <tr>
            {titles.map(getTitles)}
          </tr>
        </thead>

        <tbody>
          {[...Array(rowCount)].map((row, rowIndex) =>
            <tr
              key={rowIndex} // eslint-disable-line react/no-array-index-key
              className={(hasItems ? '' : styles.emptyTd)}>

              {
                hasItems ?
                  rows[rowIndex].map(getRow) :
                  rows.map(getRow)
              }
            </tr>
          )}
        </tbody>
      </table>
    </div>
  );
}

Table.propTypes = {
  defaultRowCount: PropTypes.number,
  defaultTitleCount: PropTypes.number,
  items: PropTypes.object,
  tableStyle: PropTypes.string
};

Table.defaultProps = {
  defaultRowCount: 8,
  defaultTitleCount: 5,
  items: {},
  tableStyle: 'full-panel'
};
