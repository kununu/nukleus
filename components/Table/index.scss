@import '~kununu-theme/scss/base/variables';

.tableContainer {
  margin-bottom: 15px;

  * {
    box-sizing: border-box;
  }

  @media (max-width: $screen-md-max) {
    overflow: scroll;
  }
}

.table {
  width: 100%;
  min-width: 1060px;
  border: 1px solid $table-border-color;

  & > thead > tr > th {
    padding: 10px 20px;
    background-color: $teal;
    border-color: $th-border-color;
    border-bottom: 0;
    text-transform: uppercase;
    font-size: $font-size-xs;
    text-align: left;

    span {
      color: $white;
    }
  }

  & > tbody > tr {
    min-height: 50px;

    &:nth-of-type(odd) {
      background-color: $table-bg-accent;
    }

    > td {
      padding: 12px 18px;
      max-width: 250px;
      overflow: hidden;
      text-overflow: ellipsis;
      font-size: $font-size-base;
      border: 1px solid $td-border-color;
    }
  }

  .emptyHeader {
    height: 40px;
    background-color: #e7e9e9;
  }

  .emptyTd {
    height: 50px;
  }

  .highlighted {
    background-color: #fcf8e3;
  }
}
