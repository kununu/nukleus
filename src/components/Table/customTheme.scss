@import '~@kununu/kununu-theme-v2/scss/base/variables';

.table {
  min-width: 0;
  width: 100%;
  overflow-x: auto;

  th {
    padding: 10px;
    vertical-align: bottom;
    font-weight: normal;
    background: $gray-base-90;
  }

  tr {
    border-bottom: 1px solid $gray-base-10;

    &:hover {
      background-color: $gray-base-10;
    }
  }

  td {
    border: 0;
    padding: 10px;
  }

  @media (min-width: $screen-sm-min) {
    td, th {
      padding: 12px 18px;
    }
  }
}

.tableHeaderContainer {
  position: relative;
  color: $white;
}

.tableSortingContainer {
  cursor: pointer;
  position: absolute;
  right: 0;
  bottom: 0;

  .tableSortingButton {
    background: none;
    color: inherit;
    border: 0;
    padding: 0;
    font: inherit;
    cursor: pointer;
    outline: inherit;
    margin-left: 2px;
    display: inline;

    .tableSortingArrowIcon {
      width: 15px;
      height: 15px;
      fill: $white;
    }
  }
}
