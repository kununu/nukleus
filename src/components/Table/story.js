import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions';
import {withOptions} from '@storybook/addon-options';
// import {action} from '@storybook/addon-actions';
// import {withNotes} from '@storybook/addon-notes';
// import {withInfo} from '@storybook/addon-info';
// import {withKnobs, text, boolean} from '@storybook/addon-knobs/react';
// import backgrounds from '@storybook/addon-backgrounds';

import Table from './index';

storiesOf('Table', module)
  .addDecorator(withOptions({...defaultOptions}))
  .add('with some emoji', () => (
    <Table
      dataRows={[
        {
          age: 11,
          color: 'red',
          name: 'Joe'
        },
        {
          age: 20,
          color: 'blue',
          name: 'Merry'
        }
      ]}
      columns={[
        {
          accessor: 'name',
          header: 'Name',
          sortable: true
        },
        {
          accessor: 'age',
          header: 'Age',
          sortable: false
        },
        {
          accessor: 'color',
          cell: val => <span>custom cell: {val}</span>,
          header: 'color',
          sortable: true
        }
      ]} />
  ));
