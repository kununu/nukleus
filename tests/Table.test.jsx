import React from 'react';
import {Link} from 'react-router-dom';
import {StaticRouter} from 'react-router';
import renderer from 'react-test-renderer';
import Table from 'Table'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies


test('Renders Table without crashing', () => {
  const component = renderer.create((
    <StaticRouter location="test" context={{}}>
      <Table
        items={{
          id: [1, 2, 3, 4, 5],
          'in hiragana': ['いち', 'に', 'さん', 'よん', 'ご'],
          'in kanji': ['一', '二', '三', '四', '五'],
          'in words': ['one', 'two', 'three', 'four', 'five'],
          link: [
            <Link to="/playground/">One</Link>,
            <Link to="/playground/">Two</Link>,
            <Link to="/playground/">Three</Link>,
            'Four',
            <Link to="/playground/">Five</Link>
          ]
        }} />
    </StaticRouter>
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders empty Table without crashing', () => {
  const component = renderer.create(<Table />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

