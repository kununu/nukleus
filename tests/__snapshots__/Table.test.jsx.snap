exports[`test Renders Table without crashing 1`] = `
<div
  className="tableContainer full-panel">
  <table
    className="table">
    <thead>
      <tr>
        <th>
          <span>
            id
          </span>
        </th>
        <th>
          <span>
            in hiragana
          </span>
        </th>
        <th>
          <span>
            in kanji
          </span>
        </th>
        <th>
          <span>
            in words
          </span>
        </th>
        <th>
          <span>
            link
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="">
        <td>
          1
        </td>
        <td>
          いち
        </td>
        <td>
          一
        </td>
        <td>
          one
        </td>
        <td>
          <a
            onClick={[Function]}
            style={Object {}}>
            One
          </a>
        </td>
      </tr>
      <tr
        className="">
        <td>
          2
        </td>
        <td>
          に
        </td>
        <td>
          二
        </td>
        <td>
          two
        </td>
        <td>
          <a
            onClick={[Function]}
            style={Object {}}>
            Two
          </a>
        </td>
      </tr>
      <tr
        className="">
        <td>
          3
        </td>
        <td>
          さん
        </td>
        <td>
          三
        </td>
        <td>
          three
        </td>
        <td>
          <a
            onClick={[Function]}
            style={Object {}}>
            Three
          </a>
        </td>
      </tr>
      <tr
        className="">
        <td>
          4
        </td>
        <td>
          よん
        </td>
        <td>
          四
        </td>
        <td>
          four
        </td>
        <td>
          Four
        </td>
      </tr>
      <tr
        className="">
        <td>
          5
        </td>
        <td>
          ご
        </td>
        <td>
          五
        </td>
        <td>
          five
        </td>
        <td>
          <a
            onClick={[Function]}
            style={Object {}}>
            Five
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`test Renders empty Table without crashing 1`] = `
<div
  className="tableContainer full-panel">
  <table
    className="table">
    <thead>
      <tr>
        <th
          className="emptyHeader" />
        <th
          className="emptyHeader" />
        <th
          className="emptyHeader" />
        <th
          className="emptyHeader" />
        <th
          className="emptyHeader" />
      </tr>
    </thead>
    <tbody>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr
        className="emptyTd">
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
